pkg/lightning/backend/local/duplicate.go

```diff
@@ -0,0 +1,637 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package local
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"sort"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	split "github.com/pingcap/br/pkg/restore"
+
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/distsql"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/ranger"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
+
+	backendkv "github.com/pingcap/br/pkg/lightning/backend/kv"
+	"github.com/pingcap/br/pkg/lightning/common"
+	"github.com/pingcap/br/pkg/lightning/log"
+	"github.com/pingcap/br/pkg/logutil"
+
+	"github.com/cockroachdb/pebble"
+	"github.com/pingcap/errors"
+	sst "github.com/pingcap/kvproto/pkg/import_sstpb"
+	kvrpc "github.com/pingcap/kvproto/pkg/kvrpcpb"
+	tikv "github.com/pingcap/kvproto/pkg/tikvpb"
+)
+
+const (
+	maxWriteBatchCount = 128
+)
+
+type DuplicateRequest struct {
+	tableID   int64
+	indexID   int64 // 0 represent it is a table request
+	start     kv.Key
+	end       kv.Key
+	indexInfo *model.IndexInfo
+}
+
+type DuplicateManager struct {
+	db                *pebble.DB
+	splitCli          split.SplitClient
+	regionConcurrency int
+	connPool          common.GRPCConns
+	tls               *common.TLS
+	ts                uint64
+	keyAdapter        KeyAdapter
+}
+
+func NewDuplicateManager(
+	db *pebble.DB,
+	splitCli split.SplitClient,
+	ts uint64,
+	tls *common.TLS,
+	regionConcurrency int) (*DuplicateManager, error) {
+	return &DuplicateManager{
+		db:                db,
+		tls:               tls,
+		regionConcurrency: regionConcurrency,
+		splitCli:          splitCli,
+		keyAdapter:        duplicateKeyAdapter{},
+		ts:                ts,
+		connPool:          common.NewGRPCConns(),
+	}, nil
+}
+
+func (manager *DuplicateManager) CollectDuplicateRowsFromTiKV(ctx context.Context, tbl table.Table) error {
+	log.L().Info("Begin collect duplicate data from remote TiKV")
+	reqs, err := buildDuplicateRequests(tbl.Meta())
+	if err != nil {
+		return err
+	}
+
+	decoder, err := backendkv.NewTableKVDecoder(tbl, &backendkv.SessionOptions{
+		SQLMode: mysql.ModeStrictAllTables,
+	})
+	if err != nil {
+		return err
+	}
+	rpcctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	g, rpcctx := errgroup.WithContext(ctx)
```

Comment:
```suggestion
	g, rpcctx := errgroup.WithContext(rpcctx)
```

Reply:
The above two lines are not required. 
```go
rpcctx, cancel := context.WithCancel(ctx)
defer cancel()
```
`errgroup` propagates `cancel` in `WithContext` and call `cancel` when `Wait` is called.

Reply:
good suggestion